ArcGISRuntimeSDKQt_CppSamples/EditData/SnapGeometryEdits/SnapGeometryEdits.cpp

```diff
@@ -0,0 +1,335 @@
+{
+  m_polylineSourceCheckedState[index] = snappingCheckedState;
+  m_polylineSourceList[index]->setEnabled(m_polylineSourceCheckedState[index]);
+
+}
+
+// Enable snapping for all the point layer snap sources
+void SnapGeometryEdits::onPointLayersEnabled()
+{
+  for (int index = 0; index < m_pointSourceList.size(); ++index)
+  {
+    pointSourceEnabledStatus(true, index);
+  }
+  emit pointSourceCheckedStateChanged();
+}
+
+// Enable snapping for all the polyline layer snap sources
+void SnapGeometryEdits::onPolylineLayersEnabled()
+{
+  for (int index = 0; index < m_polylineSourceList.size(); ++index)
+  {
+    polylineSourceEnabledStatus(true, index);
+  }
+  emit polylineSourceCheckedStateChanged();
+}
+
+// Starts the GeometryEditor using the selected geometry type
+void SnapGeometryEdits::startEditor(GeometryEditorMode geometryEditorMode)
+{
+  switch (geometryEditorMode)
+  {
+  case GeometryEditorMode::PointMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Point);
+    break;
+
+  case GeometryEditorMode::MultipointMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Multipoint);
+    break;
+
+  case GeometryEditorMode::PolylineMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Polyline);
+    break;
+
+  case GeometryEditorMode::PolygonMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Polygon);
+    break;
+
+  default:
+    break;
+  }
+  emit geometryEditorStartedChanged();
+}
+
+// Stops the GeometryEditor and append the new graphic to the graphics overlay
+void SnapGeometryEdits::stopEditor()
+{
+  const Geometry geometry = m_geometryEditor->stop();
+  emit geometryEditorStartedChanged();
+
+  if (m_editingGraphic)
+  {
+    m_editingGraphic->setGeometry(geometry);
+
+    m_editingGraphic->setVisible(true);
+    m_editingGraphic = nullptr;
+
+    return;
+  }
+
+  Symbol* geometrySymbol = nullptr;
+  switch (geometry.geometryType())
+  {
+  case GeometryType::Point:
+    geometrySymbol = m_pointSymbol;
+    break;
+
+  case GeometryType::Multipoint:
+    geometrySymbol = m_multiPointSymbol;
+    break;
+
+  case GeometryType::Polyline:
+    geometrySymbol = m_lineSymbol;
+    break;
+
+  case GeometryType::Polygon:
+    geometrySymbol = m_polygonSymbol;
+    break;
+
+  default:
+    return;
+  }
+
+  m_graphicsOverlay->graphics()->append(new Graphic(geometry, geometrySymbol, new QObject(this)));
+}
+
+// Deletes the selected element
+void SnapGeometryEdits::deleteSelection()
+{
+  m_geometryEditor->deleteSelectedElement();
+  emit canUndoChanged();
+}
+
+// Reverts the last event on the geometry editor
+void SnapGeometryEdits::editorUndo()
+{
+  m_geometryEditor->undo();
+  emit canUndoChanged();
+}
+
+void SnapGeometryEdits::displaySnapSources()
+{
+  // create lists for displaying the snap sources in snap settings
+  QList<Esri::ArcGISRuntime::SnapSourceSettings*> snapSourceList;
```

Comment:
feels like the logic of this function is a bit confusing. I believe the point of it is to get a data structure where you could create a "layer list" UI that has the name of the layer and a visibility checkbox next to it. It might be better to create a separate list model type in the sample. It would likely be more boilerplate code, but I think it would be more reliable than updating all these various members. Might be worth a chat to see if that is worth the effort or if it's overkill. cc @JamesMBallard I welcome your thoughts

Reply:
If displaying this data is a primary part of the sample experience, a list model would make sense. I don't know enough about this particular sample to know if that's overkill or not. We may have existing samples that build similar UI experiences so we should see what they do.

Reply:
I think in other samples, we have a LayerListModel coming from the API itself. In this API, we just give a QList of sources, as they  could be feature layers now, but later could include graphics overlays

Reply:
In that case, a list model of `SnapSourceSettings` would probably work and should alleviate some of the other state management/logic concerns.

@har13205 please see if that's possible. It'll be more code, but in the end it should simplify the workflow/ui management.